File: dac/backend/src/main/java/com/dremio/dac/api/Dataset.java
/*
 * Copyright (C) 2017-2018 Dremio Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.dremio.dac.api;

import java.util.List;

import org.apache.arrow.vector.types.pojo.Field;

import com.dremio.service.namespace.dataset.proto.AccelerationSettings;
import com.dremio.service.namespace.dataset.proto.RefreshMethod;
import com.dremio.service.namespace.file.FileFormat;
import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.databind.annotation.JsonSerialize;

/**
 * Dataset model
 */
@JsonIgnoreProperties(value={ "fields" }, allowGetters=true, ignoreUnknown = true)
public class Dataset implements CatalogEntity {
  /**
   * Dateset Type
   */
  public enum DatasetType {
    VIRTUAL_DATASET, PHYSICAL_DATASET
  };

  private final String id;
  private final DatasetType type;
  private final List<String> path;
  @JsonSerialize(using = DatasetFieldSerializer.class)
  private List<Field> fields;
  @JsonISODateTime
  private final Long createdAt;
  private final String tag;
  private final RefreshSettings accelerationRefreshPolicy;

  // for VDS
  private final String sql;
  private final List<String> sqlContext;

  // for promoted PDS
  private final FileFormat format;

  public Dataset(
    String id,
    DatasetType type,
    List<String> path,
    List<Field> fields,
    Long createdAt,
    String tag,
    RefreshSettings accelerationRefreshPolicy,
    String sql,
    List<String> sqlContext,
    FileFormat format) {
    this.id = id;
    this.type = type;
    this.path = path;
    this.fields = fields;
    this.createdAt = createdAt;
    this.tag = tag;
    this.accelerationRefreshPolicy = accelerationRefreshPolicy;
    this.sql = sql;
    this.sqlContext = sqlContext;
    this.format = format;
  }

  @JsonCreator
  public Dataset(
    @JsonProperty("id") String id,
    @JsonProperty("type") DatasetType type,
    @JsonProperty("path") List<String> path,
    @JsonProperty("createdAt") Long createdAt,
    @JsonProperty("tag") String tag,
    @JsonProperty("accelerationRefreshPolicy") RefreshSettings accelerationRefreshPolicy,
    @JsonProperty("sql") String sql,
    @JsonProperty("sqlContext") List<String> sqlContext,
    @JsonProperty("format") FileFormat format) {
    // we don't want to deserialize fields ever since they are immutable anyways
    this(id, type, path, null, createdAt, tag, accelerationRefreshPolicy, sql, sqlContext, format);
  }

  public String getId() {
    return id;
  }

  public DatasetType getType() {
    return type;
  }

  public List<String> getPath() {
    return path;
  }

  public List<Field> getFields() {
    return fields;
  }

  public Long getCreatedAt() {
    return createdAt;
  }

  public String getTag() {
    return tag;
  }

  public String getSql() {
    return sql;
  }

  public List<String> getSqlContext() {
    return sqlContext;
  }

  public FileFormat getFormat() {
    return format;
  }

  public RefreshSettings getAccelerationRefreshPolicy() {
    return accelerationRefreshPolicy;
  }

  /**
   * Dataset acceleration refresh settings
   */
  public static class RefreshSettings {
    private final String refreshField;
    private final Long refreshPeriodMs;
    private final Long gracePeriodMs;
    private final RefreshMethod method;

    @JsonCreator
    public RefreshSettings(
      @JsonProperty("refreshField") String refreshField,
      @JsonProperty("refreshPeriodMs") Long refreshPeriodMs,
      @JsonProperty("gracePeriodMs") Long gracePeriodMs,
      @JsonProperty("method") RefreshMethod method
    ) {
      this.refreshField = refreshField;
      this.refreshPeriodMs = refreshPeriodMs;
      this.gracePeriodMs = gracePeriodMs;
      this.method = method;
    }

    public RefreshSettings(AccelerationSettings settings) {
      refreshField = settings.getRefreshField();
      method = settings.getMethod();
      refreshPeriodMs = settings.getRefreshPeriod();
      gracePeriodMs = settings.getGracePeriod();
    }

    public String getRefreshField() {
      return refreshField;
    }

    public Long getRefreshPeriodMs() {
      return refreshPeriodMs;
    }

    public Long getGracePeriodMs() {
      return gracePeriodMs;
    }

    public RefreshMethod getMethod() {
      return method;
    }

    public AccelerationSettings toAccelerationSettings() {
      AccelerationSettings settings = new AccelerationSettings();

      settings.setRefreshPeriod(getRefreshPeriodMs());
      settings.setGracePeriod(getGracePeriodMs());
      settings.setMethod(getMethod());
      settings.setRefreshField(getRefreshField());

      return settings;
    }
  }
}

File: ui/dremio-js/src/community/catalog/Dataset.ts
/*
 * Copyright (C) 2017-2019 Dremio Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import type { SonarV3Config } from "../../_internal/types/Config.js";
import type {
  CommunityDataset,
  CommunityDatasetProperties,
} from "../../interfaces/Dataset.js";
import { DatasetCatalogReference } from "./CatalogReference.js";

import { mappedType, pathString } from "./utils.js";

export class Dataset implements CommunityDataset {
  readonly catalogReference: DatasetCatalogReference;
  readonly createdAt: Date;
  /**
   * @deprecated
   */
  readonly id: string;
  readonly fields: any;
  readonly owner: any;
  /**
   * @deprecated
   */
  readonly path: string[];
  readonly type: CommunityDatasetProperties["type"];
  // eslint-disable-next-line no-unused-private-class-members
  readonly #config: SonarV3Config;
  constructor(
    properties: CommunityDatasetProperties & {
      catalogReference: DatasetCatalogReference;
    },
    config: SonarV3Config,
  ) {
    this.catalogReference = properties.catalogReference;
    this.createdAt = properties.createdAt;
    this.id = properties.id;
    this.fields = properties.fields;
    this.owner = properties.owner;
    this.path = properties.path;
    this.type = properties.type;
    this.#config = config;
  }

  /**
   * @deprecated
   */
  get name(): string {
    return this.path[this.path.length - 1]!;
  }

  /**
   * @deprecated
   */
  pathString = pathString(() => this.path);

  /**
   * @deprecated
   */
  get referenceType() {
    return this.type;
  }

  static fromResource(datasetEntity: any, config: SonarV3Config): Dataset {
    return new Dataset(
      {
        catalogReference: new DatasetCatalogReference({
          id: datasetEntity.id,
          path: datasetEntity.path,
          type: (mappedType as any)[datasetEntity.type],
        }),
        createdAt: new Date(datasetEntity.createdAt),
        fields: datasetEntity.fields,
        id: datasetEntity.id,
        owner: datasetEntity.owner
          ? {
              id: datasetEntity.owner.ownerId,
              type: datasetEntity.owner.ownerType,
            }
          : undefined,
        path: datasetEntity.path,
        type: (mappedType as any)[datasetEntity.type],
      },
      config,
    );
  }
}
